components/turret.py

```diff
@@ -107,110 +81,70 @@ def scan(self, heading):
-        if -self.CLOSED_LOOP_ERROR < current_error < self.CLOSED_LOOP_ERROR:
-            self.ticks_within_threshold += 1
-        else:
-            self.ticks_within_threshold = 0
-        return self.ticks_within_threshold > self.TICKS_TO_SETTLE
-
-    def _do_slewing(self):
-        if self._slewing_is_finished():
-            self.current_state = self.IDLE
-            self.target_count = 0
-
-    ################
-    #
-    # The remainder of this file, as well as the FINDING_INDEX state and all
-    # the relevant references, can be removed once we no longer explicitly
-    # do indexing.
-    #
-    ################
-
-    # Seek for 200ms at first, before reversing and doubling
-    STARTING_MAX_TICKS = 10
-
-    # The indexing does not use the encoder at all right now. It just counts
-    # the times it's called, incrementing (or decrementing) a tick count each
-    # time. This method resets the counting mechanism.
-    def _reset_ticks(self) -> None:
-        self.motor_speed = 0.2  # Flips when seeking, so not a constant
-        self.seeking = False
-        self.tick_count = 0
-        self.max_ticks = self.STARTING_MAX_TICKS
-        self.max_ticks_factor = -2
-        self.tick_increment = 1
-
-    # Find the nearest index and reset the encoder
-    def run_indexing(self) -> None:
-        self._reset_ticks()
-        self.current_state = self.FINDING_INDEX
-
-    def _do_indexing(self):
-        # Are we there yet? If so, greb the encoder value, stop the motor,
-        # and stop seeking
-        if self._index_found() == Index.CENTRE:
-            self.baseline_azimuth = self.motor.getSelectedSensorPosition(0)
-            self.baseline_is_provisional = False
-            self.motor.stopMotor()
-            self._reset_ticks()
-            self.current_state = self.IDLE
-            # self.logger.info("Found the sensor")
-            return
-
-        # If we haven't started yet, start seeking
-        if not self.seeking:
-            self.motor.set(ctre.ControlMode.PercentOutput, self.motor_speed)
-            self.tick_count = 0
-            self.max_ticks = self.STARTING_MAX_TICKS
-            self.seeking = True
-            # self.logger.info("starting to seek")
-            return
-
-        self.tick_count = self.tick_count + self.tick_increment
-        if self.tick_count == self.max_ticks:
-            self.motor.stopMotor()
-            self.motor_speed = -self.motor_speed
-            self.motor.set(ctre.ControlMode.PercentOutput, self.motor_speed)
-            self.tick_increment = self.tick_increment * -1
-            self.max_ticks = self.max_ticks * self.max_ticks_factor
-            return
-
-        # Currently running and haven't hit the limit nor found an index.
-        # Just keep the motor running
-        self.motor.set(ctre.ControlMode.PercentOutput, self.motor_speed)
+            _reset_encoder(0)
+        # TODO: Repeat for other index marks
+
+    def _reset_encoder(self, counts) -> None:
```

Comment:
The hall-effect sensors have a non-trivial angle over which they fire, so the exact encoder value will depend on which direction we approached from and either add or subtract half the sensor width in counts to set the correct absolute angle of the turret in counts. One reason to isolate the conversion of the index we found into counts in _index_to_count is to encapsulate dealing with this.
Also, naming the function _check_for_index is slightly misleading. It does more than check. It checks and then changes the state of the motor encoder if it finds one. I separated these on purpose to make the code clearer and easier to understand and change later. Side effects make code harder to understand and to change.

Reply:
That is a good point. Do we have an idea of what the angle of activation is (and whether it is significant)?
To find the midpoint of the activation is going to require some more logic to spot the rising then falling edge. It's not impossible to do.
I take the point about the method name. I'm not sure how having a lookup function for the particular index to encoder counts helps solve the above problem. I don't think I follow what you are proposing.

Reply:
Or are you proposing we measure the sensor activation width ahead of time and then use that fixed value in the calculation? This would work, but we would still have to make each call to reset the encoder a one-shot to avoid multiple resets as we traverse the hall effect sensor.

Reply:
Yes, that's what I'm proposing. I see what you mean about avoiding multiple resets during traversal.

This actually adds weight to my proposal to separate determining when we hit an index, which now has to worry about returning true only once during traversal, and what we do about having found it, which involves changing motor encoders and resetting targets, taking into account the traversal width. My point is simply that even though these happen together, they are different, so making them different functions makes each one simpler and easier to change.

Reply:
Oh, and Nicky and I measured the hall-effect sensor activation width to be about 7mm (3.5 mm on either side). We should repeat that measurement more precisely in the lab, but the angle would be 3.5mm as a fraction of the circumference of the turret at the magnet. It will be small, but I don't think it will be negligible.

Reply:
I have pushed a commit that looks for the first activation and accounts for the sensor width.

Reply:
Looks perfect.